includes/tracking/class-pixel.php

```diff
@@ -260,37 +265,91 @@ public static function log_pixel_url( $url, $post_id, $tracking_id, $email_tag )
 		);
 	}
 
+	/**
+	 * Sanitize and track an item from the log.
+	 *
+	 * @param string $item The item to sanitize and track.
+	 */
+	public static function sanitize_and_track_item( $item ) {
+		$item = trim( $item ); // Trim trailing line break.
+		$item = explode( '|', $item );
+		if ( 3 !== count( $item ) ) {
+			return;
+		}
+
+		// Values must be sanitized as they are stored in the logs without sanitization.
+		$newsletter_id = isset( $item[0] ) ? intval( $item[0] ) : 0;
+		$tracking_id   = isset( $item[1] ) ? \sanitize_text_field( $item[1] ) : 0;
+		$email_address = isset( $item[2] ) ? \sanitize_email( $item[2] ) : '';
+		if ( ! $newsletter_id || ! $tracking_id || ! $email_address ) {
+			return;
+		}
+
+		self::track_seen( $newsletter_id, $tracking_id, $email_address );
+	}
+
 	/**
 	 * Process logs.
 	 */
 	public static function process_logs() {
 		$current_log_file = \get_option( 'newspack_newsletters_tracking_pixel_log_file' );
 
 		if ( $current_log_file && file_exists( $current_log_file ) ) {
-			// Read the tracking data from the log file.
-			$data = file_get_contents( $current_log_file ); // phpcs:ignore WordPressVIPMinimum.Performance.FetchingRemoteData.FileGetContentsUnknown
-
-			// Process the tracking data.
-			$tracking_data = explode( PHP_EOL, $data );
-			foreach ( $tracking_data as $item ) {
-				if ( ! $item ) {
-					continue;
+			// Read the tracking data from the log file. Process in batches to avoid memory issues.
+			$handle    = fopen( $current_log_file, 'r' ); // phpcs:ignore WordPress.WP.AlternativeFunctions.file_system_operations_fopen
+			$file_end  = false;
+			$pos       = 0;
+
+			while ( ! $file_end ) {
+				rewind( $handle );
+				$file_size = filesize( $current_log_file );
+				$chunk_end = false;
+				$lines     = 0;
+
+				// Process a chunk of lines from the file.
+				while ( ! $chunk_end ) {
+					// If we've processed a full chunk.
+					if ( $lines >= self::MAX_LINES - 1 ) {
+						$chunk_end = true;
+					}
+
+					// Process the tracking data.
+					$item = fgets( $handle );
+
+					// If we've reached the end of the chunk, stop the loop.
+					if ( ! $item ) {
+						$chunk_end = true;
+					}
+
+					self::sanitize_and_track_item( $item );
+					$lines++;
 				}
-				$item = explode( '|', $item );
-				if ( 3 !== count( $item ) ) {
-					continue;
+
+				// Get the current position in the file after processing the chunk.
+				$pos = ftell( $handle );
+
+				// If we've reached the end of the file, stop the loop.
+				if ( $pos >= $file_size ) {
```

Comment:
In my testing, this condition is never met, and the `$file_size` is never updated from its initial value. 

Testing steps:
1. Hardcode the value of `$current_log_file` to `<file-name>`
2. Generate a 150-lines logfile: `printf '1|2|hey@email.com\n%.0s' {1..150} > <file-name>`
3. Trigger the logfile processing: `wp eval "\Newspack_Newsletters\Tracking\Pixel::process_logs();"` 

The eval function never exits, and then the `$file_size` is logged, it's never updated. The file is cleared, though. If the `process_logs` execution is halted, then the method called again, the file is removed since the file size is initially `0`.

Reply:
Whoops, I must have lost a `break;` statement in there somewhere. ac4f352 fixes this